File: sdks/go/sip/commitment.go
package sip

import (
	"crypto/rand"
	"crypto/sha256"
	"errors"
	"fmt"
	"math/big"

	"github.com/decred/dcrd/dcrec/secp256k1/v4"
)

// Domain separation tag for H generation
const hDomain = "SIP-PEDERSEN-GENERATOR-H-v1"

var (
	// curveOrder is the secp256k1 curve order
	curveOrder = secp256k1.S256().N

	// generatorG is the base generator
	generatorG = secp256k1.Generator()

	// generatorH is the independent generator (NUMS)
	generatorH *secp256k1.JacobianPoint
)

func init() {
	// Generate H using NUMS method
	generatorH = generateH()
}

// generateH generates the independent generator H using NUMS method.
func generateH() *secp256k1.JacobianPoint {
	for counter := 0; counter < 256; counter++ {
		// Create candidate x-coordinate
		input := fmt.Sprintf("%s:%d", hDomain, counter)
		hash := sha256.Sum256([]byte(input))

		// Try to create a point from this x-coordinate (with even y)
		pointBytes := make([]byte, 33)
		pointBytes[0] = 0x02 // Compressed, even y
		copy(pointBytes[1:], hash[:])

		pubKey, err := secp256k1.ParsePubKey(pointBytes)
		if err == nil {
			var result secp256k1.JacobianPoint
			pubKey.AsJacobian(&result)
			return &result
		}
	}

	panic("failed to generate H point - this should never happen")
}

// Commit creates a Pedersen commitment to a value.
//
// C = v*G + r*H
//
// Where:
//   - v = value (the amount being committed)
//   - r = blinding factor (random, keeps value hidden)
//   - G = base generator
//   - H = independent generator (NUMS)
func Commit(value uint64) (HexString, HexString, error) {
	// Generate random blinding factor
	blinding := make([]byte, 32)
	_, err := rand.Read(blinding)
	if err != nil {
		return "", "", fmt.Errorf("failed to generate blinding: %w", err)
	}

	return CommitWithBlinding(value, blinding)
}

// CommitWithBlinding creates a Pedersen commitment with a specific blinding factor.
func CommitWithBlinding(value uint64, blinding []byte) (HexString, HexString, error) {
	if len(blinding) != 32 {
		return "", "", errors.New("blinding must be 32 bytes")
	}

	// Convert to scalars
	vScalar := new(secp256k1.ModNScalar)
	vScalar.SetInt(uint32(value))

	rScalar := new(secp256k1.ModNScalar)
	rScalar.SetByteSlice(blinding)

	if rScalar.IsZero() {
		return "", "", errors.New("zero blinding scalar - investigate RNG")
	}

	// C = v*G + r*H
	var vG, rH, commitment secp256k1.JacobianPoint

	if value == 0 {
		// Only blinding contributes: C = r*H
		secp256k1.ScalarMultNonConst(rScalar, generatorH, &commitment)
	} else {
		// v*G
		secp256k1.ScalarBaseMultNonConst(vScalar, &vG)
		// r*H
		secp256k1.ScalarMultNonConst(rScalar, generatorH, &rH)
		// C = v*G + r*H
		secp256k1.AddNonConst(&vG, &rH, &commitment)
	}

	commitment.ToAffine()
	pubKey := secp256k1.NewPublicKey(&commitment.X, &commitment.Y)
	commitmentBytes := pubKey.SerializeCompressed()

	return BytesToHex(commitmentBytes), BytesToHex(blinding), nil
}

// VerifyOpening verifies that a commitment opens to a specific value.
//
// Recomputes C' = v*G + r*H and checks if C' == C
func VerifyOpening(commitment HexString, value uint64, blinding HexString) (bool, error) {
	commitmentBytes, err := HexToBytes(commitment)
	if err != nil {
		return false, fmt.Errorf("invalid commitment: %w", err)
	}

	blindingBytes, err := HexToBytes(blinding)
	if err != nil {
		return false, fmt.Errorf("invalid blinding: %w", err)
	}

	// Parse commitment point
	cPubKey, err := secp256k1.ParsePubKey(commitmentBytes)
	if err != nil {
		return false, fmt.Errorf("invalid commitment point: %w", err)
	}

	// Recompute expected commitment
	vScalar := new(secp256k1.ModNScalar)
	vScalar.SetInt(uint32(value))

	rScalar := new(secp256k1.ModNScalar)
	rScalar.SetByteSlice(blindingBytes)

	var expected secp256k1.JacobianPoint
	if value == 0 {
		secp256k1.ScalarMultNonConst(rScalar, generatorH, &expected)
	} else {
		var vG, rH secp256k1.JacobianPoint
		secp256k1.ScalarBaseMultNonConst(vScalar, &vG)
		secp256k1.ScalarMultNonConst(rScalar, generatorH, &rH)
		secp256k1.AddNonConst(&vG, &rH, &expected)
	}

	expected.ToAffine()
	expectedPubKey := secp256k1.NewPublicKey(&expected.X, &expected.Y)

	return cPubKey.IsEqual(expectedPubKey), nil
}

// CommitZero creates a commitment to zero with a specific blinding factor.
//
// C = 0*G + r*H = r*H
func CommitZero(blinding []byte) (HexString, HexString, error) {
	return CommitWithBlinding(0, blinding)
}

// AddCommitments adds two commitments homomorphically.
//
// C1 + C2 = (v1*G + r1*H) + (v2*G + r2*H) = (v1+v2)*G + (r1+r2)*H
func AddCommitments(c1, c2 HexString) (HexString, error) {
	c1Bytes, err := HexToBytes(c1)
	if err != nil {
		return "", fmt.Errorf("invalid c1: %w", err)
	}

	c2Bytes, err := HexToBytes(c2)
	if err != nil {
		return "", fmt.Errorf("invalid c2: %w", err)
	}

	pubKey1, err := secp256k1.ParsePubKey(c1Bytes)
	if err != nil {
		return "", fmt.Errorf("invalid commitment c1: %w", err)
	}

	pubKey2, err := secp256k1.ParsePubKey(c2Bytes)
	if err != nil {
		return "", fmt.Errorf("invalid commitment c2: %w", err)
	}

	var point1, point2, sum secp256k1.JacobianPoint
	pubKey1.AsJacobian(&point1)
	pubKey2.AsJacobian(&point2)

	secp256k1.AddNonConst(&point1, &point2, &sum)
	sum.ToAffine()

	resultPubKey := secp256k1.NewPublicKey(&sum.X, &sum.Y)
	return BytesToHex(resultPubKey.SerializeCompressed()), nil
}

// SubtractCommitments subtracts two commitments homomorphically.
//
// C1 - C2 = (v1-v2)*G + (r1-r2)*H
func SubtractCommitments(c1, c2 HexString) (HexString, error) {
	c1Bytes, err := HexToBytes(c1)
	if err != nil {
		return "", fmt.Errorf("invalid c1: %w", err)
	}

	c2Bytes, err := HexToBytes(c2)
	if err != nil {
		return "", fmt.Errorf("invalid c2: %w", err)
	}

	pubKey1, err := secp256k1.ParsePubKey(c1Bytes)
	if err != nil {
		return "", fmt.Errorf("invalid commitment c1: %w", err)
	}

	pubKey2, err := secp256k1.ParsePubKey(c2Bytes)
	if err != nil {
		return "", fmt.Errorf("invalid commitment c2: %w", err)
	}

	var point1, point2, negPoint2, diff secp256k1.JacobianPoint
	pubKey1.AsJacobian(&point1)
	pubKey2.AsJacobian(&point2)

	// Negate point2
	negPoint2 = point2
	negPoint2.Y.Negate(1)
	negPoint2.Y.Normalize()

	secp256k1.AddNonConst(&point1, &negPoint2, &diff)
	diff.ToAffine()

	resultPubKey := secp256k1.NewPublicKey(&diff.X, &diff.Y)
	return BytesToHex(resultPubKey.SerializeCompressed()), nil
}

// AddBlindings adds blinding factors (for use with homomorphic addition).
func AddBlindings(b1, b2 HexString) (HexString, error) {
	b1Bytes, err := HexToBytes(b1)
	if err != nil {
		return "", fmt.Errorf("invalid b1: %w", err)
	}

	b2Bytes, err := HexToBytes(b2)
	if err != nil {
		return "", fmt.Errorf("invalid b2: %w", err)
	}

	s1 := new(secp256k1.ModNScalar)
	s1.SetByteSlice(b1Bytes)

	s2 := new(secp256k1.ModNScalar)
	s2.SetByteSlice(b2Bytes)

	sum := s1.Add(s2)
	return BytesToHex(sum.Bytes()[:]), nil
}

// SubtractBlindings subtracts blinding factors (for use with homomorphic subtraction).
func SubtractBlindings(b1, b2 HexString) (HexString, error) {
	b1Bytes, err := HexToBytes(b1)
	if err != nil {
		return "", fmt.Errorf("invalid b1: %w", err)
	}

	b2Bytes, err := HexToBytes(b2)
	if err != nil {
		return "", fmt.Errorf("invalid b2: %w", err)
	}

	s1 := new(secp256k1.ModNScalar)
	s1.SetByteSlice(b1Bytes)

	s2 := new(secp256k1.ModNScalar)
	s2.SetByteSlice(b2Bytes)

	s2.Negate()
	diff := s1.Add(s2)
	return BytesToHex(diff.Bytes()[:]), nil
}

// GenerateBlinding generates a random blinding factor.
func GenerateBlinding() (HexString, error) {
	bytes := make([]byte, 32)
	_, err := rand.Read(bytes)
	if err != nil {
		return "", fmt.Errorf("failed to generate blinding: %w", err)
	}
	return BytesToHex(bytes), nil
}

// GetGenerators returns the generators for ZK proof integration.
func GetGenerators() Generators {
	gPoint := secp256k1.Generator()
	var gJac secp256k1.JacobianPoint
	gPoint.AsJacobian(&gJac)

	return Generators{
		Gx: BytesToHex(gJac.X.Bytes()[:]),
		Gy: BytesToHex(gJac.Y.Bytes()[:]),
		Hx: BytesToHex(generatorH.X.Bytes()[:]),
		Hy: BytesToHex(generatorH.Y.Bytes()[:]),
	}
}

File: sdks/rust/src/commitment.rs
//! Pedersen Commitment Implementation for SIP Protocol.
//!
//! Cryptographically secure Pedersen commitments on secp256k1.
//!
//! # Security Properties
//!
//! - **Hiding (Computational)**: Cannot determine value from commitment
//! - **Binding (Computational)**: Cannot open commitment to different value
//! - **Homomorphic**: C(v1) + C(v2) = C(v1 + v2) when blindings sum
//!
//! # Generator H Construction
//!
//! H is constructed using "nothing-up-my-sleeve" (NUMS) method to ensure
//! nobody knows the discrete log of H w.r.t. G.

use k256::{
    elliptic_curve::{
        group::GroupEncoding,
        sec1::{FromEncodedPoint, ToEncodedPoint},
        Field, PrimeField,
    },
    AffinePoint, ProjectivePoint, Scalar,
};
use rand::RngCore;
use sha2::{Digest, Sha256};

use crate::crypto::{bytes_to_hex, hex_to_bytes};
use crate::error::{Error, Result};
use crate::types::HexString;

/// Domain separation tag for H generation
const H_DOMAIN: &str = "SIP-PEDERSEN-GENERATOR-H-v1";

/// The base generator G (secp256k1)
fn get_generator_g() -> ProjectivePoint {
    ProjectivePoint::GENERATOR
}

/// Generate the independent generator H using NUMS method.
fn generate_h() -> ProjectivePoint {
    for counter in 0..256 {
        // Create candidate x-coordinate
        let input = format!("{}:{}", H_DOMAIN, counter);
        let mut hasher = Sha256::new();
        hasher.update(input.as_bytes());
        let hash = hasher.finalize();

        // Try to create a point from this x-coordinate (with even y)
        let mut point_bytes = [0u8; 33];
        point_bytes[0] = 0x02; // Compressed, even y
        point_bytes[1..].copy_from_slice(&hash);

        if let Ok(point) = AffinePoint::from_bytes(&point_bytes.into()) {
            let proj = ProjectivePoint::from(point);
            if !proj.is_identity().into() && proj != ProjectivePoint::GENERATOR {
                return proj;
            }
        }
    }

    panic!("Failed to generate H point - this should never happen");
}

lazy_static::lazy_static! {
    static ref H: ProjectivePoint = generate_h();
}

/// Create a Pedersen commitment to a value.
///
/// C = v*G + r*H
///
/// Where:
/// - v = value (the amount being committed)
/// - r = blinding factor (random, keeps value hidden)
/// - G = base generator
/// - H = independent generator (NUMS)
///
/// # Arguments
///
/// * `value` - The value to commit to
///
/// # Returns
///
/// Tuple of (commitment, blinding) as hex strings
///
/// # Example
///
/// ```rust
/// use sip_protocol::commit;
///
/// let (commitment, blinding) = commit(100).unwrap();
/// ```
pub fn commit(value: u64) -> Result<(HexString, HexString)> {
    // Generate random blinding factor
    let mut blinding_bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut blinding_bytes);

    commit_with_blinding(value, &blinding_bytes)
}

/// Create a Pedersen commitment with a specific blinding factor.
pub fn commit_with_blinding(value: u64, blinding: &[u8]) -> Result<(HexString, HexString)> {
    if blinding.len() != 32 {
        return Err(Error::CryptoError("Blinding must be 32 bytes".to_string()));
    }

    let g = get_generator_g();

    // Convert value and blinding to scalars
    let v_scalar = Scalar::from(value);
    let r_scalar = Scalar::from_repr_vartime((*blinding).into())
        .ok_or_else(|| Error::CryptoError("Invalid blinding scalar".to_string()))?;

    if r_scalar.is_zero().into() {
        return Err(Error::CryptoError(
            "Zero blinding scalar - investigate RNG".to_string(),
        ));
    }

    // C = v*G + r*H
    let commitment = if value == 0 {
        // Only blinding contributes: C = r*H
        *H * r_scalar
    } else {
        // Normal case: C = v*G + r*H
        g * v_scalar + *H * r_scalar
    };

    let commitment_bytes = commitment.to_affine().to_bytes();

    Ok((
        bytes_to_hex(&commitment_bytes),
        bytes_to_hex(blinding),
    ))
}

/// Verify that a commitment opens to a specific value.
///
/// Recomputes C' = v*G + r*H and checks if C' == C
///
/// # Arguments
///
/// * `commitment` - The commitment point to verify
/// * `value` - The claimed value
/// * `blinding` - The blinding factor used
///
/// # Returns
///
/// true if the commitment opens correctly
pub fn verify_opening(commitment: &str, value: u64, blinding: &str) -> Result<bool> {
    let commitment_bytes = hex_to_bytes(commitment)?;
    let blinding_bytes = hex_to_bytes(blinding)?;

    let g = get_generator_g();

    // Parse commitment point
    let c_point = AffinePoint::from_bytes(commitment_bytes.as_slice().into())
        .map(ProjectivePoint::from)
        .map_err(|_| Error::InvalidPublicKey("Invalid commitment point".to_string()))?;

    // Recompute expected commitment
    let v_scalar = Scalar::from(value);
    let r_scalar = Scalar::from_repr_vartime(blinding_bytes.as_slice().try_into().unwrap())
        .ok_or_else(|| Error::CryptoError("Invalid blinding scalar".to_string()))?;

    let expected = if value == 0 {
        *H * r_scalar
    } else {
        g * v_scalar + *H * r_scalar
    };

    Ok(c_point == expected)
}

/// Create a commitment to zero with a specific blinding factor.
///
/// C = 0*G + r*H = r*H
pub fn commit_zero(blinding: &[u8]) -> Result<(HexString, HexString)> {
    commit_with_blinding(0, blinding)
}

/// Add two commitments homomorphically.
///
/// C1 + C2 = (v1*G + r1*H) + (v2*G + r2*H) = (v1+v2)*G + (r1+r2)*H
pub fn add_commitments(c1: &str, c2: &str) -> Result<HexString> {
    let c1_bytes = hex_to_bytes(c1)?;
    let c2_bytes = hex_to_bytes(c2)?;

    let point1 = AffinePoint::from_bytes(c1_bytes.as_slice().into())
        .map(ProjectivePoint::from)
        .map_err(|_| Error::InvalidPublicKey("Invalid commitment c1".to_string()))?;

    let point2 = AffinePoint::from_bytes(c2_bytes.as_slice().into())
        .map(ProjectivePoint::from)
        .map_err(|_| Error::InvalidPublicKey("Invalid commitment c2".to_string()))?;

    let sum = point1 + point2;
    Ok(bytes_to_hex(&sum.to_affine().to_bytes()))
}

/// Subtract two commitments homomorphically.
///
/// C1 - C2 = (v1-v2)*G + (r1-r2)*H
pub fn subtract_commitments(c1: &str, c2: &str) -> Result<HexString> {
    let c1_bytes = hex_to_bytes(c1)?;
    let c2_bytes = hex_to_bytes(c2)?;

    let point1 = AffinePoint::from_bytes(c1_bytes.as_slice().into())
        .map(ProjectivePoint::from)
        .map_err(|_| Error::InvalidPublicKey("Invalid commitment c1".to_string()))?;

    let point2 = AffinePoint::from_bytes(c2_bytes.as_slice().into())
        .map(ProjectivePoint::from)
        .map_err(|_| Error::InvalidPublicKey("Invalid commitment c2".to_string()))?;

    let diff = point1 - point2;
    Ok(bytes_to_hex(&diff.to_affine().to_bytes()))
}

/// Add blinding factors (for use with homomorphic addition).
pub fn add_blindings(b1: &str, b2: &str) -> Result<HexString> {
    let b1_bytes = hex_to_bytes(b1)?;
    let b2_bytes = hex_to_bytes(b2)?;

    let s1 = Scalar::from_repr_vartime(b1_bytes.as_slice().try_into().unwrap())
        .ok_or_else(|| Error::CryptoError("Invalid blinding b1".to_string()))?;
    let s2 = Scalar::from_repr_vartime(b2_bytes.as_slice().try_into().unwrap())
        .ok_or_else(|| Error::CryptoError("Invalid blinding b2".to_string()))?;

    let sum = s1 + s2;
    Ok(bytes_to_hex(&sum.to_bytes()))
}

/// Subtract blinding factors (for use with homomorphic subtraction).
pub fn subtract_blindings(b1: &str, b2: &str) -> Result<HexString> {
    let b1_bytes = hex_to_bytes(b1)?;
    let b2_bytes = hex_to_bytes(b2)?;

    let s1 = Scalar::from_repr_vartime(b1_bytes.as_slice().try_into().unwrap())
        .ok_or_else(|| Error::CryptoError("Invalid blinding b1".to_string()))?;
    let s2 = Scalar::from_repr_vartime(b2_bytes.as_slice().try_into().unwrap())
        .ok_or_else(|| Error::CryptoError("Invalid blinding b2".to_string()))?;

    let diff = s1 - s2;
    Ok(bytes_to_hex(&diff.to_bytes()))
}

/// Generate a random blinding factor.
pub fn generate_blinding() -> HexString {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    bytes_to_hex(&bytes)
}

/// Get the generators for ZK proof integration.
pub fn get_generators() -> (HexString, HexString, HexString, HexString) {
    let g = get_generator_g().to_affine();
    let h = H.to_affine();

    let g_encoded = g.to_encoded_point(false);
    let h_encoded = h.to_encoded_point(false);

    (
        bytes_to_hex(g_encoded.x().unwrap()),
        bytes_to_hex(g_encoded.y().unwrap()),
        bytes_to_hex(h_encoded.x().unwrap()),
        bytes_to_hex(h_encoded.y().unwrap()),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_commit_and_verify() {
        let (commitment, blinding) = commit(100).unwrap();
        assert!(verify_opening(&commitment, 100, &blinding).unwrap());
        assert!(!verify_opening(&commitment, 101, &blinding).unwrap());
    }

    #[test]
    fn test_homomorphic_addition() {
        let (c1, b1) = commit(100).unwrap();
        let (c2, b2) = commit(50).unwrap();

        let c_sum = add_commitments(&c1, &c2).unwrap();
        let b_sum = add_blindings(&b1, &b2).unwrap();

        assert!(verify_opening(&c_sum, 150, &b_sum).unwrap());
    }
}
